hypothesis-python/src/hypothesis/internal/conjecture/utils.py

```diff
@@ -126,18 +127,18 @@ def check_sample(values):
                 'want to sample slices.  Sampling a multi-dimensional '
                 'array will be an error in a future version of Hypothesis.'
             ).format(ndim=values.ndim, shape=values.shape))
-    elif not isinstance(values, (OrderedDict, Sequence, enum.EnumMeta)):
+    elif not isinstance(values, _SEQUENCE_TYPES):
         note_deprecation(
-            ('Cannot sample from %r, not a sequence.  ' % (values,)) +
-            'Hypothesis goes to some length to ensure that sampling an '
-            'element from a collection (with `sampled_from` or `choices`) is '
-            'replayable and can be minimised.  To replay a saved example, '
-            'the sampled values must have the same iteration order on every '
-            'run - ruling out sets, dicts, etc due to hash randomisation.  '
-            'Most cases can simply use `sorted(values)`, but mixed types or '
-            'special values such as math.nan require careful handling - and '
-            'note that when simplifying an example, Hypothesis treats '
-            'earlier values as simpler.')
+            'Cannot sample from {values}, not an ordered collection. '
+            'Hypothesis goes to some length to ensure that the {strategy} '
+            'strategy has stable results between runs. To replay a saved '
+            'example, the sampled values must have the same iteration order '
+            'on every run - ruling out sets, dicts, etc due to hash '
+            'randomisation. Most cases can simply use `sorted(values)`, but '
+            'mixed types or special values such as math.nan require careful '
+            'handling - and note that when simplifying an example, '
+            'Hypothesis treats earlier values as simpler.'.format(
```

Comment:
I wonder if we want a different deprecation message for permutations? Fiddling with the sort order for `permutations` will affect the shrink order, but not necessarily in this way – we shrink towards the original order, don’t we?

Reply:
I have no suggestions off the top of my head as to what the message should be, but you're right.

Reply:
> Cannot sample from {values}, not an ordered collection. Hypothesis tries to ensure that the permutations strategy has stable results between runs. To replay a saved example, the sampled values must have the same iteration order on every run - ruling out sets, dicts, etc due to hash randomisation. This strategy treats permutations that are closer to the original order of the collection as simpler.

?

Reply:
This is outside the scope of this PR, but if you do want to have different checks and error messages per `SearchStrategy`, it might make sense to refactor `check_sample` into an instance method so it can be overridden per strategy.

Reply:
Happy to leave it out of scope for this PR :smile: